pennylane/_queuing.py

```diff
@@ -0,0 +1,197 @@
+    with the current architecture. Therefore, there is a need to use a high level
+    object that holds information about the relationship between quantum gates and
+    a quantum node.
+
+    The :class:`~.QueuingContext` class realizes this by providing access to
+    the current QNode.  Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
+
+    The ``QueuingContext`` class both acts as the abstract base class for all
+    classes that expose a queue for Operations (so-called contexts), as well as the
+    interface to said queues. The active contexts contain maximally one QNode and
+    an arbitrary number of other contexts like the :class:`~.OperationRecorder`.
+    """
+
+    # TODO: update docstring
+
+    _active_contexts = []
+    """The list of contexts that are currently active."""
+
+    def __enter__(self):
+        """Adds this instance to the global list of active contexts.
+
+        Returns:
+            QueuingContext: This instance
+        """
+        QueuingContext._active_contexts.append(self)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Remove this instance from the global list of active contexts.
+        """
+        QueuingContext._active_contexts.remove(self)
+
+    @abc.abstractmethod
+    def _append(self, obj):
+        """Append an object to this QueuingContext instance.
+
+        Args:
+            obj: The object to be appended
+        """
+
+    @classmethod
+    def append(cls, obj):
+        """Append an object to the queue(s).
+
+        Args:
+            obj: the object to be appended
+        """
+        for context in cls._active_contexts:
+            context._append(obj)  # pylint: disable=protected-access
+
+    @abc.abstractmethod
+    def _remove(self, obj):
+        """Remove an object from this QueuingContext instance.
+
+        Args:
+            obj: the object to be removed
+        """
+
+    @classmethod
+    def remove(cls, obj):
+        """Remove an object from the queue(s) if it is in the queue(s).
+
+        Args:
+            obj: the object to be removed
+        """
+        for context in cls._active_contexts:
+            # We use the duck-typing approach to assume that the underlying remove
+            # behaves like list.remove and throws a ValueError if the operator
+            # is not in the list
+            try:
+                context._remove(obj)  # pylint: disable=protected-access
+            except ValueError:
+                pass
+
+
+class Queue(QueuingContext):
+    """Lightweight class that maintains a basic queue of operations and pre/post-processing steps
+    for representing quantum circuits."""
+
+    def __init__(self):
+        self.queue = []
+
+    def _append(self, operator):
+        self.queue.append(operator)
+
+    def _remove(self, obj):
+        self.queue.remove(obj)
+
+
+class OperationRecorder(Queue):
```

Comment:
Other than changing the parent class, and the requisite change in method names/signatures, nothing else was done here